[new module] Adding module azure_rm_sqldatabase

zikalino opened the pull request:
##### SUMMARY
Adding support for Azure SQL Databases

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
azure_rm_sql_databases

##### ANSIBLE VERSION


##### ADDITIONAL INFORMATION
This module depends on azure_rm_sql_servers

Commit:
added module azure-rm-sql-databases

Commit:
rename databases -> database

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_databases.py:
```diff
@@ -0,0 +1,443 @@
+#!/usr/bin/python
```
Seems like you copied the original but didn't remove it :)

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_database.py:
```diff
@@ -0,0 +1,442 @@
+---
+module: azure_rm_sql_database
+version_added: "2.5"
+short_description: Manage an Databases.
```
We don't put `.` on the short_description.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_database.py:
```diff
@@ -0,0 +1,442 @@
+    location:
+        description:
+            - Resource location.
+        required: True
```
Like with `azure_sql_server`, can we not require this and derive the location from the resource group if not specified?

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_database.py:
```diff
@@ -0,0 +1,442 @@
+        required: False
+    create_mode:
+        description:
+            - "Specifies the mode of database creation.\n\nDefault: regular database creation.\n\nCopy: creates a database as a copy of an existing database.
```
Let's not copy and paste the values from the documentation and do something like.

```
create_mode:
  description:
  - Specifies the mode of the database creation.
  - C(Default) is a regular database creation
  - C(Copy) creates a database as a copy of an existing database, I(source_database_id) must be set in this mode.
  - C(OnlineSecondary) creates a database as a readable secondary replica, I(source_database_id) must be set in this mode.
  - ... continue for each option and state what needs to be set
  default: Default

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_database.py:
```diff
@@ -0,0 +1,442 @@
+               ecovery point resource ID.\n\nCopy, NonReadableSecondary, OnlineSecondary and RestoreLongTermRetentionBackup are not supported for DataWarehou
+               se edition. Possible values include: 'Copy', 'Default', 'NonReadableSecondary', 'OnlineSecondary', 'PointInTimeRestore', 'Recovery', 'Restore'
+               , 'RestoreLongTermRetentionBackup'"
+        required: False
```
`required: False` not needed.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_database.py:
```diff
@@ -0,0 +1,442 @@
+        required: False
+    source_database_id:
+        description:
+            - "Conditional. If createMode is Copy, NonReadableSecondary, OnlineSecondary, PointInTimeRestore, Recovery, or Restore, then this value is requir
```
Split up as an entry per sentence instead of one long line, e.g.

```
source_database_id:
  description:
  - Specifies the resource ID of the source database.
  - If I(create_mode) is C(Copy), C(NonReadableSecondary), ...., then this must be set.
  - If I(create_mode) is C(NonReadableSecondary) or C(OnlineSecondary), the name of the source db must be the same as the new db.
```

Maybe even make it clearer whether this is the name of the database or the id, or both?

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_database.py:
```diff
@@ -0,0 +1,442 @@
+'''
+
+EXAMPLES = '''
+      - name: Create (or update) Sql
```
A few things

* Keep the examples left aligned (remove the leading 4 spaces)
* Stop using jinja2 block `{{}}` in the examples so people can see what the actual may look like
* Provide more than 1 example, usually we have a basic with only the parameters required, then 1 for each major scenario, like 1 for the main create_mode options and so on

jborean93 reviewed (changes requested):
I haven't tested this one yet but a lot of the comments I had for your azure_rm_sql_server also apply this this one here when it comes to the code. One area I think this needs to be improved on is the documentation. I started putting a few comments but they mostly apply to all the options and it all balls down to;

* Don't copy and paste from docs as the option names don't match up to the Ansible side
* Put the text after a `.` on a new line like in the examples I gave you
* When talking about behaviour for different options, try and put them on their own sentance.
* Expand the examples as there seems to be quite a few different ways of doing things

We should also look at changing the return values to only return what the user may not already know but we can look at that after we get the documentation updated and the tests expanded.

Commit:
udpated databases module as requested in review

jborean93 commented:
One more name change as well, the common format is to use `azure_rm_< resource name without _ >`, in this case would be `azure_rm_sqldatabase` and the same thing would apply to the sql server module as well.

Commit:
renamed database

Commit:
replaced has_key by in

Commit:
removed irrelevant fields from response

zikalino commented:
ready_for_review

zikalino commented:
fixed several major issues:
- renaming, etc.
- idempotence support
- added update / idempotence tests
I may have missed some smaller issues.
Please recheck if still any important changes needed

Additional note - I still plan to add following adjustments in this PR:
- flatten state in response (like in Contatiner Registry module)
- add default values
- flatten parameters

Commit:
flattened response and a few other changes

jborean93 commented on test/integration/targets/azure_rm_sqldatabase/aliases:
```diff
@@ -0,0 +1,2 @@
+cloud/azure
```
Are you able to update this so it runs on the CI

jborean93 commented on test/integration/targets/azure_rm_sqldatabase/tasks/main.yml:
```diff
@@ -0,0 +1,50 @@
+    location: eastus
+    admin_username: mylogin
+    admin_password: Testpasswordxyz12!
+- name: Create instance of SQL Database
```
Are you able to add a check mode and idempotent task for the create, update and delete tasks

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+version_added: "2.5"
+short_description: Manage SQL Database instance
+description:
+    - Create, update and delete instance of SQL Database
```
Full stop

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+        description:
+            - The name of the database to be operated on (updated or created).
+        required: True
+    tags:
```
The extends documentation fragments `azure_tags` adds this option

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+            - Resource tags.
+    location:
+        description:
+            - Resource location.
```
What is the default value if not set, from the resource group?

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+            - Resource location.
+    collation:
+        description:
+            - The collation of the database. If createMode is not Default, this value is ignored.
```
s/createMode/I(create_mode)/.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+            - The collation of the database. If createMode is not Default, this value is ignored.
+    create_mode:
+        description:
+            - "Specifies the mode of database creation.\n\nDefault: regular database creation.\n\nCopy: creates a database as a copy of an existing database.
```
This really needs to be split up instead of just one long paragraph that seems copied from the Azure Docs. Set each options under the options key and Create a new description category for each option.

```
create_mode:
  description:
  - Specifies the mode of data creation.
  - When C(default) means a regular database creation.
  - When C(copy) will create a database as a copy of an existing data. I(source_database_id) must be specified.
  - When C().....
  default: default
  options:
  - default
  - copy
  - non_readable_secondary
  - online_secondary
  - point_in_time_restore
  - recovery
  - restore
  - restore_long_term_retention_backup
```

I would also convert the options to use snake_cake and not CamelCase as that is what Ansible usually does.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+
+EXAMPLES = '''
+  - name: Create (or update) SQL Database
+    azure_rm_sqldatabase:
```
Please use a real world example, also having more than 1 to handle the different create_mode scenarios would be great.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+            ),
+            tags=dict(
+                type='dict',
+                required=False
```
Not needed

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+            ),
+            location=dict(
+                type='str',
+                required=False
```
Not needed

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+                required=False
+            ),
+            source_database_deletion_date=dict(
+                type='datetime',
```
Not a valid type, see https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L855-L868 for actual options.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+    def exec_module(self, **kwargs):
+        """Main module execution method"""
+
+        for key in list(self.module_arg_spec.keys()) + ['tags']:
```
You're not doing anything with these values so instead of putting in all these if statements just change it to

```
for key in list(self.module_arg_spec.keys()) + ['tags']:
    setattr(self, key, kwargs[key])
```

You would need to make sure you set `self.key = None` in the `__init__` constructor above but this is the common way Azure modules are done.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+        old_response = None
+        results = dict()
+
+        self.mgmt_client = self.get_mgmt_svc_client(SqlManagementClient,
```
As we talked about in the WG yesterday, please set the API version in the module and pass it along here.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+                                                    base_url=self._cloud_environment.endpoints.resource_manager)
+
+        try:
+            resource_group = self.get_resource_group(self.resource_group)
```
No need for the try/except, `get_resource_group` already handles this.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+        except CloudError:
+            self.fail('resource group {0} not found'.format(self.resource_group))
+
+        if not ("location" in self.parameters):
```
You can just change this to

```
if not self.location:
    self.location = resource_group.location
```

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+        self.log("Creating / Updating the SQL Database instance {0}".format(self.name))
+
+        try:
+            response = self.mgmt_client.databases.create_or_update(self.resource_group,
```
Use kwargs instead of positional args

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+        '''
+        self.log("Deleting the SQL Database instance {0}".format(self.name))
+        try:
+            response = self.mgmt_client.databases.delete(self.resource_group,
```
kwargs

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+        self.log("Checking if the SQL Database instance {0} is present".format(self.name))
+        found = False
+        try:
+            response = self.mgmt_client.databases.get(self.resource_group,
```
kwargs

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+            self.log("Need to Create / Update the SQL Database instance")
+
+            if self.check_mode:
+                return self.results
```
This would need `results['changed'] = True` in the create scenario. We really need a way to checking if we need to update the database instead of doing the operation and checking if it changed there as well.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+            if not old_response:
+                self.results['changed'] = True
+            else:
+                self.results['changed'] = old_response.__ne__(response)
```
This isn't really how we usually go update operations in Ansible. Is there way of checking if a change needs to occur and then making or rather than making a request and seeing if it was changed. The benefit of the former approach is that when running in check mode you can see if a change needs to happen while in this scenario you wouldn't know.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqldatabase.py:
```diff
@@ -0,0 +1,483 @@
+                self.results['changed'] = True
+            else:
+                self.results['changed'] = old_response.__ne__(response)
+            self.results.update(response)
```
Don't do this as you would need to pop all the results below. Just set `results['id'] = response....` for each value you want to return to make it simpler.

jborean93 commented on test/integration/targets/azure_rm_sqldatabase/tasks/main.yml:
```diff
@@ -0,0 +1,50 @@
+  assert:
+    that:
+      - output.changed
+      - output.status == 'Online'
```
Would be good to check that some of the return values like `id` is defined, like so

```
- name: Assert the resource instance is created
  assert:
  that:
  - output.changed
  - output.status == 'Online'
  - output.id is defined
  - output.database_id is defined
```

jborean93 reviewed (changes requested):
Have put some comments, the main thing would be to add some more tests and fix up the documentation so it isn't just a copy and paste of some Azure docs and more related to the module itself.
